src/python/pants/rules/core/test.py

```diff
@@ -173,24 +181,24 @@ class AddressAndDebugRequest:
 
 @rule
 async def coordinator_of_tests(
-  target: HydratedTarget,
-  union_membership: UnionMembership,
-  address_origin_map: AddressOriginMap
+  target_with_origin: HydratedTargetWithOrigin, union_membership: UnionMembership,
 ) -> AddressAndTestResult:
+  target = target_with_origin.target
+  adaptor_with_origin = TargetAdaptorWithOrigin.create(
+    adaptor=target.adaptor, origin=target_with_origin.origin
+  )
```

Comment:
This is key to how the union works. `adaptor_with_origin` will have a specific type like `PythonTestsAdaptorWithOrigin`. That then gets passed to `await Get[TestResult](TestTarget, adaptor_with_origin)` and the union does its magic.